Add documentation related to removal of source and recovery source in k-NN performance tuning section (#7359)
opensearch-project/documentation-website: _search-plugins/knn/performance-tuning.md

## Indexing performance tuning

> Take any of the following steps to improve indexing performance, especially when you plan to index a large number of vectors at once:

Reviewer: ```suggestion
Take any of the following steps to improve indexing performance, especially when you plan to index a large number of vectors at once.
```

       }
   }
   ```

> Make sure to reenable `refresh_interval` after indexing finishes.

Reviewer: ```suggestion
Make sure to reenable `refresh_interval` after indexing is complete.
```

### Disable replicas (no OpenSearch replica shard)

>    Set replicas to `0` to prevent duplicate construction of native library indexes in both primary and replica shards. When you enable replicas after indexing finishes, the serialized native library indexes are directly copied. If you have no replicas, losing nodes might cause data loss, so it's important that the data lives elsewhere so this initial load can be retried in case of an issue.

Reviewer: ```suggestion
   Set replicas to `0` to prevent duplicate construction of native library indexes in both primary and replica shards. When you enable replicas after indexing completes, the serialized native library indexes are copied directly. If you have no replicas, losing nodes might cause data loss, so it's important that the data be stored elsewhere so that this initial load can be retried in the event of an issue.
```

### Increase the number of indexing threads

> If the hardware you choose has multiple cores, you can allow multiple threads in native library index construction by speeding up the indexing process. Determine the number of threads to allot with the [knn.algo_param.index_thread_qty]({{site.url}}{{site.baseurl}}/search-plugins/knn/settings#cluster-settings) setting.

Reviewer: ```suggestion
If your hardware has multiple cores, you can allow multiple threads in native library index construction by speeding up the indexing process. Determine the number of threads to allot with the [knn.algo_param.index_thread_qty]({{site.url}}{{site.baseurl}}/search-plugins/knn/settings#cluster-settings) setting.
```

> Keep an eye on CPU utilization and choose the correct number of threads. Because native library index construction is costly, having multiple threads can cause additional CPU load.

Reviewer: ```suggestion
Monitor CPU utilization and choose the correct number of threads. Because native library index construction is costly, having multiple threads can cause additional CPU load.
```

> Keep an eye on CPU utilization and choose the correct number of threads. Because native library index construction is costly, having multiple threads can cause additional CPU load.

Reviewer: Nit: Maybe a more precise verb than "having"?

> ### (Expert-level) Disable storing of vector fields in source

Reviewer: "Disable vector field storage in the source"?

> ### (Expert-level) Disable storing of vector fields in source

Reviewer 2: ```suggestion
### (Expert-level) Disable vector field storage in the source field
```

> The `_source` field contains the original JSON document body that was passed at index time. This field is not indexed and is not searchable but is stored so that it can be returned when executing fethc requests such as `get` and `search`. When using vector fields within the source, you can remove the vector field to save up disk space, as shown in the following example where the `location` vector is excluded:

Reviewer: ```suggestion
The `_source` field contains the original JSON document body that was passed at index time. This field is not indexed and is not searchable but is stored so that it can be returned when executing fetch requests such as `get` and `search`. When using vector fields within the source, you can remove the vector field to save disk space, as shown in the following example where the `location` vector is excluded:
```

  ```json
  PUT /<index_name>/_mappings
  {
      "_source": {
        "excludes": ["location"]
      },
      "properties": {
          "location": {
              "type": "knn_vector",
              "dimension": 2,
              "method": {
                  "name": "hnsw",
                  "space_type": "l2",
                  "engine": "faiss"
              }
          }
      }
  }
  ```

  

> After disabling the `_source` for the vector field API like update, update_by_query, reindex will not as expected because OpenSearch has no way to know what was the vector.

Reviewer 2: How is `update_by_query` related to the removal of the vector field and the reindex operation? 

```suggestion
After disabling the `_source` for the vector field API like update, update_by_query, reindex will not as expected because OpenSearch has no way to know what was the vector.
```

Author: The side effects is coming from removal of a field from _source. In this case because that field is vector field.

Reviewer 2: So `update` and `reindex` operations involving the disabled field will not work?

Reviewer 2: Trying to understand what the side effects you are referring to are.

Author: pushed a commit.

Reviewer 2: That still doesn't clarify the relation between the "update_by_query" operation, the missing vector, and reindexing. 

Would the following rewording be accurate? 

"After disabling the `_source` vector field, vector API operations, such as `update_by_query` or `reindex`, will not operate since OpenSearch can no longer find the vector".

Author: The behavior is true for any field and not related to vector field. We say the same thing here in this PR too: [_field-types/metadata-fields/source.md](https://github.com/opensearch-project/documentation-website/pull/6933/files#diff-9a227ae77d563b0eb8cb9cdc1f5bacb86f7bb000d6768162c34848aa11aa0a97)

I am just referring the same thing here because the behavior is same.

> Disabling the `_source` field can cause certain featues to become not supported, such as `update`, `update_by_query`, and `reindex` APIs and the ability to debug queries or aggregations by using the original document at index time.

Reviewer: ```suggestion
Disabling the `_source` field can cause certain features to become unavailable, such as the `update`, `update_by_query`, and `reindex` APIs and the ability to debug queries or aggregations by using the original document at index time.
```

> In OpenSearch 2.15 or great, you can further improve the indexing speed and reduce disk space, by removing remove vector field from `_recovery_source`, as shown in the following example:

Reviewer: ```suggestion
In OpenSearch 2.15 or later, you can further improve indexing speed and reduce disk space by removing the vector field from the `_recovery_source`, as shown in the following example:
```

> In OpenSearch 2.15 or great, you can further improve the indexing speed and reduce disk space, by removing remove vector field from `_recovery_source`, as shown in the following example:

Reviewer: Should "vector" be in code font?

Reviewer 2: No, since its referring to vector fields generally.

  ```json
  PUT /<index_name>/_mappings
  {
      "_source": {
        "excludes": ["location"],
        "recovery_source_excludes": ["location"]
      },
      "properties": {
          "location": {
              "type": "knn_vector",
              "dimension": 2,
              "method": {
                  "name": "hnsw",
                  "space_type": "l2",
                  "engine": "faiss"
              }
          }
      }
  }
  ```

> This is an expert-level setting. Disabling the `_recovery_source` may lead to failures during peer to peer recovery. Before disabling the `_recovery_source`, check with your OpenSearch cluster admin to know see if your cluster performs regular flushes before starting the peer to peer recovery of shards before disabling the `_recovery_source`.  

Reviewer: ```suggestion
This is an expert-level setting. Disabling the `_recovery_source` may lead to failures during peer-to-peer recovery. Before disabling the `_recovery_source`, check with your OpenSearch cluster admin to determine whether your cluster performs regular flushes before starting the peer-to-peer recovery of shards before disabling the `_recovery_source`.  
```

   The warmup API operation loads all native library indexes for all shards (primary and replica) for the specified indexes into the cache, so there's no penalty to load native library indexes during initial searches.

> This API operation only loads the segments of the indexes it ***sees*** into the cache. If a merge or refresh operation finishes after the API runs, or if you add new documents, you need to rerun the API to load those native library indexes into memory.

Reviewer: The first sentence here is a bit awkward because the operation can't literally "see" anything, including index segments. Can we revise for clarity?

Reviewer 2: How about "active" indexes?

> This API operation only loads the segments of the indexes it ***sees*** into the cache. If a merge or refresh operation finishes after the API runs, or if you add new documents, you need to rerun the API to load those native library indexes into memory.

Reviewer 2: ```suggestion
This API operation only loads the segments of active indexes into the cache. If a merge or refresh operation finishes after the API runs, or if you add new documents, you need to rerun the API to load those native library indexes into memory.
```
